File: test/Zorglub.Tests.CSharp/Time/Simple/CalendarMathTests.cs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2020 Narvalo.Org. All rights reserved.

namespace Zorglub.Time.Simple;

public static partial class CalendarMathTests
{
    private static readonly FauxCalendarMath s_Math = new(GregorianCalendar.Instance);

    private static readonly CalendarDate s_Date = new(1, 1, 1, Cuid.Gregorian);
    private static readonly CalendarMonth s_Month = new(1, 1, Cuid.Gregorian);

    private static readonly CalendarDate s_InvalidDate = new(1, 1, 1, Cuid.Julian);
    private static readonly CalendarMonth s_InvalidMonth = new(1, 1, Cuid.Julian);

    private static readonly CalendarDate s_OtherDate = new(2000, 1, 1, Cuid.Gregorian);
    private static readonly CalendarMonth s_OtherMonth = new(2000, 1, Cuid.Gregorian);
}

public partial class CalendarMathTests
{
    [Fact]
    public static void Constructor_WithInvalidCalendar() =>
        Assert.ThrowsAnexn("calendar", () => new FauxCalendarMath(null!));

    [Fact]
    public static void Constructor_WithCalendar()
    {
        var chr = GregorianCalendar.Instance;
        // Act
        var arithmetic = new FauxCalendarMath(GregorianCalendar.Instance);
        // Assert
        Assert.Equal(chr.Id, arithmetic.Cuid);
    }

    [Fact]
    public static void AddAdjustment_Prop() =>
        Assert.Equal(AddAdjustment.EndOfMonth, s_Math.AddAdjustment);

    [Fact]
    public static void AddYears_WithInvalidCalendar()
    {
        Assert.Throws<ArgumentException>("month", () => s_Math.AddYears(s_InvalidMonth, 1));
        Assert.Throws<ArgumentException>("date", () => s_Math.AddYears(s_InvalidDate, 1));
    }

    [Fact]
    public static void AddYears()
    {
        Assert.False(s_Math.AddYearsCoreWasCalled);
        _ = s_Math.AddYears(s_Date, 1000);
        Assert.True(s_Math.AddYearsCoreWasCalled);

        Assert.False(s_Math.AddYearsCore1WasCalled);
        _ = s_Math.AddYears(s_Month, 1000);
        Assert.True(s_Math.AddYearsCore1WasCalled);

    }

    [Fact]
    public static void AddMonths_WithInvalidCalendar()
    {
        Assert.Throws<ArgumentException>("month", () => s_Math.AddMonths(s_InvalidMonth, 1));
        Assert.Throws<ArgumentException>("date", () => s_Math.AddMonths(s_InvalidDate, 1));
    }

    [Fact]
    public static void AddMonths()
    {
        Assert.False(s_Math.AddMonthsCoreWasCalled);
        _ = s_Math.AddMonths(s_Date, 1000);
        Assert.True(s_Math.AddMonthsCoreWasCalled);

        Assert.False(s_Math.AddMonthsCore1WasCalled);
        _ = s_Math.AddMonths(s_Month, 1000);
        Assert.True(s_Math.AddMonthsCore1WasCalled);
    }

    [Fact]
    public static void CountYearsBetween_WithInvalidCalendar()
    {
        Assert.Throws<ArgumentException>("start",
            () => s_Math.CountYearsBetween(s_InvalidDate, s_Date));
        Assert.Throws<ArgumentException>("end",
            () => s_Math.CountYearsBetween(s_Date, s_InvalidDate));
    }

    [Fact]
    public static void CountYearsBetween()
    {
        Assert.False(s_Math.CountYearsBetweenCoreWasCalled);
        _ = s_Math.CountYearsBetween(s_Date, s_OtherDate);
        Assert.True(s_Math.CountYearsBetweenCoreWasCalled);
    }

    [Fact]
    public static void CountMonthsBetween_WithInvalidCalendar()
    {
        Assert.Throws<ArgumentException>("start",
            () => s_Math.CountMonthsBetween(s_InvalidMonth, s_Month));
        Assert.Throws<ArgumentException>("end",
            () => s_Math.CountMonthsBetween(s_Month, s_InvalidMonth));
        Assert.Throws<ArgumentException>("start",
            () => s_Math.CountMonthsBetween(s_InvalidDate, s_Date));
        Assert.Throws<ArgumentException>("end",
            () => s_Math.CountMonthsBetween(s_Date, s_InvalidDate));
    }

    [Fact]
    public static void CountMonthsBetween()
    {
        Assert.False(s_Math.CountMonthsBetweenCoreWasCalled);
        _ = s_Math.CountMonthsBetween(s_Date, s_OtherDate);
        Assert.True(s_Math.CountMonthsBetweenCoreWasCalled);

        Assert.False(s_Math.CountMonthsBetweenCore1WasCalled);
        _ = s_Math.CountMonthsBetween(s_Month, s_OtherMonth);
        Assert.True(s_Math.CountMonthsBetweenCore1WasCalled);
    }
}

File: test/Zorglub.Tests/Tests/Simple/CalendarMathTests.fs
﻿// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) Tran Ngoc Bich. All rights reserved.

module Zorglub.Tests.Simple.CalendarMathTests

open Zorglub.Testing

open Zorglub.Time
open Zorglub.Time.Core
open Zorglub.Time.Simple

open Xunit

module Prelude =
    [<Fact>]
    let ``Constructor throws for null calendar`` () =
        nullExn "calendar" (fun () -> new FauxCalendarMath(null))

    [<Fact>]
    let ``RegularMath constructor throws for non-regular schema`` () =
        let sch = new FauxSystemSchema()
        let chr = new FauxSystemCalendar(sch)

        argExn "calendar" (fun () -> new RegularMath(chr))

    //
    // Properties
    //

    [<Fact>]
    let ``Property AdditionRuleset`` () =
        let ruleset = AdditionRuleset(
            DateRule    = AdditionRule.Overspill,
            OrdinalRule = AdditionRule.Exact,
            MonthRule   = AdditionRule.Overflow)
        let math = new FauxCalendarMath(ruleset)

        math.AdditionRuleset === ruleset

    [<Fact>]
    let ``Property Cuid`` () =
        let chr = SimpleCalendar.Gregorian
        let math = new FauxCalendarMath(chr)

        math.Cuid === chr.Id

module Factories =
    [<Fact>]
    let ``Create() & CreateDefault() -> RegularMath when Regular12`` () =
        let sch = FauxSystemSchema.Regular12
        let chr = new FauxSystemCalendar(sch)
        let ruleset = new AdditionRuleset()
        let math1 = CalendarMath.CreateDefault(chr)
        let math2 = CalendarMath.Create(chr, ruleset)

        math1 |> is<RegularMath>
        math2 |> is<RegularMath>

    [<Fact>]
    let ``Create() & CreateDefault() -> RegularMath when Regular13`` () =
        let sch = FauxSystemSchema.Regular13
        let chr = new FauxSystemCalendar(sch)
        let ruleset = new AdditionRuleset()
        let math1 = CalendarMath.CreateDefault(chr)
        let math2 = CalendarMath.Create(chr, ruleset)

        math1 |> is<RegularMath>
        math2 |> is<RegularMath>

    [<Fact>]
    let ``Create() & CreateDefault() -> RegularMath when Regular14`` () =
        let sch = FauxSystemSchema.Regular14
        let chr = new FauxSystemCalendar(sch)
        let ruleset = new AdditionRuleset()
        let math1 = CalendarMath.CreateDefault(chr)
        let math2 = CalendarMath.Create(chr, ruleset)

        math1 |> is<RegularMath>
        math2 |> is<RegularMath>

    [<Fact>]
    let ``Create() & CreateDefault() -> PlainMath when not regular`` () =
        let sch = new FauxSystemSchema()
        let chr = new FauxSystemCalendar(sch)
        let ruleset = new AdditionRuleset()
        let math1 = CalendarMath.CreateDefault(chr)
        let math2 = CalendarMath.Create(chr, ruleset)

        math1 |> is<PlainMath>
        math2 |> is<PlainMath>

    [<Fact>]
    let ``Create()`` () =
        let sch = new FauxSystemSchema()
        let chr = new FauxSystemCalendar(sch)
        let ruleset = new AdditionRuleset(
            DateRule    = AdditionRule.Overspill,
            OrdinalRule = AdditionRule.Exact,
            MonthRule   = AdditionRule.Overflow)
        let math = CalendarMath.Create(chr, ruleset)

        math |> is<PowerMath>
        math.AdditionRuleset === ruleset

module Validation =
    let private chr = SimpleCalendar.Gregorian
    let private math = new FauxCalendarMath(chr)

    let private date = new CalendarDate(1, 1, 1, chr.Id)
    let private ordinalDate = new OrdinalDate(1, 1, chr.Id)
    let private month = new CalendarMonth(1, 1, chr.Id)

    let private invalidDate = new CalendarDate(1, 1, 1, Cuid.Julian)
    let private invalidOrdinalDate = new OrdinalDate(1, 1, Cuid.Julian)
    let private invalidMonth = new CalendarMonth(1, 1, Cuid.Julian)

    //
    // CalendarDate
    //

    [<Fact>]
    let ``AddYears(invalid date) throws`` () =
        argExn "date" (fun () -> math.AddYears(invalidDate, 1))

    [<Fact>]
    let ``AddMonths(invalid date) throws`` () =
        argExn "date" (fun () -> math.AddMonths(invalidDate, 1))

    [<Fact>]
    let ``CountYearsBetween(invalid dates) throws`` () =
        argExn "end" (fun () -> math.CountYearsBetween(date, invalidDate))
        argExn "start" (fun () -> math.CountYearsBetween(invalidDate, date))

    [<Fact>]
    let ``CountMonthsBetween(invalid dates) throws`` () =
        argExn "end" (fun () -> math.CountMonthsBetween(date, invalidDate))
        argExn "start" (fun () -> math.CountMonthsBetween(invalidDate, date))

    //
    // OrdinalDate
    //

    [<Fact>]
    let ``AddYears(invalid ordinal date) throws`` () =
        argExn "date" (fun () -> math.AddYears(invalidOrdinalDate, 1))

    [<Fact>]
    let ``CountYearsBetween(invalid ordinal dates) throws`` () =
        argExn "end" (fun () -> math.CountYearsBetween(ordinalDate, invalidOrdinalDate))
        argExn "start" (fun () -> math.CountYearsBetween(invalidOrdinalDate, ordinalDate))

    //
    // CalendarMonth
    //

    [<Fact>]
    let ``AddYears(invalid month) throws`` () =
        argExn "month" (fun () -> math.AddYears(invalidMonth, 1))

    [<Fact>]
    let ``CountYearsBetween(invalid month) throws`` () =
        argExn "end" (fun () -> math.CountYearsBetween(month, invalidMonth))
        argExn "start" (fun () -> math.CountYearsBetween(invalidMonth, month))

module Core =
    let private chr = SimpleCalendar.Gregorian
    let private math = new FauxCalendarMath(chr)

    let private date = new CalendarDate(1, 1, 1, chr.Id)
    let private ordinalDate = new OrdinalDate(1, 1, chr.Id)
    let private month = new CalendarMonth(1, 1, chr.Id)

    [<Fact>]
    let ``AddYears(date) calls AddYearsCore()`` () =
        math.AddYearsCoreDateWasCalled |> nok
        math.AddYears(date, 1000)      |> ignore
        math.AddYearsCoreDateWasCalled |> ok

    [<Fact>]
    let ``AddMonths(date) calls AddMonthsCore()`` () =
        math.AddMonthsCoreDateWasCalled |> nok
        math.AddMonths(date, 1000)      |> ignore
        math.AddMonthsCoreDateWasCalled |> ok

    [<Fact>]
    let ``AddYears(ordinal date) calls AddYearsCore()`` () =
        math.AddYearsCoreOrdinalDateWasCalled |> nok
        math.AddYears(ordinalDate, 1000)      |> ignore
        math.AddYearsCoreOrdinalDateWasCalled |> ok

    [<Fact>]
    let ``AddYears(month) calls AddYearsCore()`` () =
        math.AddYearsCoreMonthWasCalled |> nok
        math.AddYears(month, 1000)      |> ignore
        math.AddYearsCoreMonthWasCalled |> ok
